---
title: "Analysis Assignment – Pipeline Development"
author: "Samuel Perini"
date: "2/5/2021"
output: html_notebook
---

The assignment was composed of three main parts: __read coverage, variant calling and analytical performance__. For each part, I report the questions that were asked, the strategy that I have used to answer them and the results that were generated. 

## (1) Read coverage

In Next Generation Sequencing (NGS) studies, the term “coverage” describes a relation between sequence reads and a reference. A reference can be a whole genome, a specific chromosome or targeted genomic regions.  
The first question of the assignment was:  
__What can you say about the quality of the read coverage in the sample?__

In order to calculate the read coverage, I have used an interval of chromosome 19 of the human genome hg19 as a reference because the DNA sequences of the assignment were aligned to a selected portion of chromosome 19. The genomic interval of chr19 starts at position 60,004 and ends at 14,992,498, for a total length of 3,498,177 bases. This information was retrieved from the BAM file via:

```{bash interval_chr19, eval=FALSE}
samtools view ../Files_needed_for_task/chr19.bam | awk '{print $3 "\t" $4 "\t" $4+length($10)-1}' > ../Read_coverage/chr19_interval.txt
sort -nk2 ../Read_coverage/chr19_interval.txt | awk 'NR==1{print "min:"$2} END{print "max:"$3}'
```

The percetage of the chr19 interval that was covered can be calculated as:

$\left ( \frac{\text{No. of base covered}}{\text{chr19 interval}} \right) \times 100$

After calculating the number of bases with zero read coverage (564; see code below), it is possible to obtain the number of base covered $(3,498,177 - 564 = 3,497,613)$ and the percentage of covarage $(3,497,613 / 3,498,177) \times 100 = 99.98\%$

```{bash depth}
samtools depth ../Files_needed_for_task/chr19.bam > 1_read_coverage/chr19_depth.txt

# The first line of the file chr19_cov_redundancy.txt is the number of bases with zero read coverage
cut -f 3 1_read_coverage/chr19_depth.txt | sort | uniq -c > 1_read_coverage/chr19_cov_redundancy.txt
```


```{r perc_cov}
# READ DEPTH DATA GENERATED BY SAMTOOLS DEPTH
dep <- read.table(file = "1_read_coverage/chr19_depth.txt")
colnames(dep) <- c("CHROM", "POS", "DEPTH")

```



```{r coverageB}

```

## (2) Variant calling



## (3) Analytical performance

