rm(list = ls())
library(ggplot2)

# READ DEPTH DATA GENERATED BY SAMTOOLS DEPTH
dep <- read.table(file = "Bioinformatics_pipeline_development_task/Read_coverage/chr19_depth.txt")
head(dep)
tail(dep)
colnames(dep) <- c("CHROM", "POS", "DEPTH")
range(dep$POS)
table(dep$DEPTH>0)
head(dep[dep$DEPTH==0,])
sum(dep$DEPTH)
dep[dep$DEPTH==1034,]

# PLOT DEPTH ACROSS ENTIRE CHROMOSOME 19
range(dep$DEPTH)
plot(dep$DEPTH,
     type='h',
     ylim=c(0,1100),
     col='blue',
     ylab='Coverage',
     xlab='Position',
     xaxt='n',
     main='Coverage at chr19'
)

# GENERATE X-AXIS
l <- length(dep$POS)
axis(1,
     at = seq(from = 0, to = l, length.out = 18),
     labels = seq(from = head(dep$POS,n=1)-1,
                  to=tail(dep$POS, n=1),
                  by = l/4
     )
)

# READ TARGET FILE
tar <- read.table(file = "Files_needed_for_task/target_regions.bed")
head(tar)
colnames(tar) <- c("CHROM", "START", "END")
range(tar$START)
range(tar$END)

#GENOME FRACTION COVERAGE
fra <- read.table("Read_coverage/raw_data_qualimapReport/genome_fraction_coverage.txt")
head(fra)
sum(fra$V2)
