rm(list = ls())
library(ggplot2)

# Total number of bases of the chr19 region
length(60004:14992498)

# Read coverage redundancy table
redu <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_cov_redundancy.txt")
redu <- redu[order(redu$V2), ]
head(redu)

# Percentage of non-covered bases
(redu[1,1]/length(60004:14992498))*100

# Percentage of bases covered by at least one read
100-((redu[1,1]/length(60004:14992498))*100)

# Mean coverage


# READ DEPTH DATA GENERATED BY SAMTOOLS DEPTH
dep <- read.table(file = "Bioinformatics_pipeline_development_task/Read_coverage/chr19_depth.txt")
head(dep)
tail(dep)
colnames(dep) <- c("CHROM", "POS", "DEPTH")
range(dep$POS)

table(dep$DEPTH>0)
3497613/length(min(dep$POS):max(dep$POS))

head(dep[dep$DEPTH==0,])
sum(dep$DEPTH)
dep[dep$DEPTH==max(dep$DEPTH),]

deeptT <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_coverage.txt")
head(deeptT)
range(deeptT$V2)
range(deeptT$V3)
table(deeptT$V4>0)
249777/length(min(dep$POS):max(dep$POS))
mean(deeptT$V4)
deeptT[deeptT$V4==max(deeptT$V4),]
deeptT[6900:6950,]

# PLOT DEPTH ACROSS ENTIRE CHROMOSOME 19
range(dep$DEPTH)
plot(dep$DEPTH,
     type='h',
     ylim=c(0,1100),
     col='blue',
     ylab='Coverage',
     xlab='Position',
     xaxt='n',
     main='Coverage at chr19'
)

# GENERATE X-AXIS
l <- length(dep$POS)
axis(1,
     at = seq(from = 0, to = l, length.out = 18),
     labels = seq(from = head(dep$POS,n=1)-1,
                  to=tail(dep$POS, n=1),
                  by = l/4
     )
)

mean(dep$DEPTH)
bedgr <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_coverage.bedgraph")
head(bedgr)
mean(bedgr$V4)

deps <- dep[sample(x = 1:nrow(dep), size = 1000, replace = FALSE), ]
head(deps)
ggplot(data = deps) +
   geom_segment(aes(x = POS, y = 0, xend = POS, yend = DEPTH))

# READ TARGET FILE
tar <- read.table(file = "Files_needed_for_task/target_regions.bed")
head(tar)
colnames(tar) <- c("CHROM", "START", "END")
range(tar$START)
range(tar$END)

#GENOME FRACTION COVERAGE
fra <- read.table("Read_coverage/raw_data_qualimapReport/genome_fraction_coverage.txt")
head(fra)
sum(fra$V2)
