rm(list = ls())
library(ggplot2)

# Total number of bases of the chr19 region
length(60004:14992498)

# Read coverage redundancy table
redu <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_cov_redundancy.txt")
redu <- redu[order(redu$V2), ]
head(redu)

# Percentage of non-covered bases
(redu[1,1]/length(60004:14992498))*100

# Percentage of bases covered by at least one read
100-((redu[1,1]/length(60004:14992498))*100)

# Plot coverage
redu$prop <- redu$V1/sum(redu$V1)
sum(redu$prop)
head(redu)
max(redu$V2)
brks <- c(0:10, 15, seq(20,100,length.out = 9), 125, seq(150, 350,by = 50))
red_bin <- redu[redu$V2 %in% brks, ]

head(red_bin)
ggplot(data = red_bin[-1,], aes(x = as.factor(V2), y = prop)) +
   geom_histogram(stat = "identity") +
   # scale_x_continuous(breaks=c(1,2,3,4,5,10,30), limits = c(0, 31)) +
   # scale_x_continuous(breaks=c(1,2,3,4,5,10,30,100,300,1000), trans="log1p", expand=c(0,0)) +
   theme_bw() +
   labs(x = "Coverage (# reads per bp)", y = "Fraction of bases") +
   theme(axis.title = element_text(size = 16), axis.text = element_text(size = 11)) +
   annotate("text", y = 0.038, x = 25, label="Mean = 14X", hjust=1, vjust=1, size = 6)
   # scale_x_continuous(breaks = c(1, seq(10, 50, by = 10)), limits = c(0, 51), labels = c(1, seq(10, 50, by = 10)))
   
head(redu)
redu$cums <- cumsum(redu$prop)
redu$icums <- round(c(redu$cums[1], 1-redu$cums[-1]), 4)

ggplot(data = redu) +
   # geom_line(aes(x = V2, y = cums)) +
   geom_line(aes(x = V2, y = icums))

x <- rnorm(12)
Fn <- ecdf(x)
Fn     # a *function*
Fn(x)  # returns the percentiles for x

deepT <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_coverage.txt")
head(deepT)
ggplot(deepT, aes(x = V4, y = 1 - ..y..)) +
   stat_ecdf(size = 1, pad = TRUE) +
   geom_segment(aes(x = 40, y = -Inf, xend = 40, yend = 0.1), linetype = "dashed") +
   geom_segment(aes(x = -Inf, y = 0.1, xend = 40, yend = 0.1), linetype = "dashed") +
   theme_bw() +
   labs(x = "Coverage (# reads per bp)", y = "Fraction of bases >= coverage") +
   theme(axis.title = element_text(size = 16), axis.text = element_text(size = 11)) +
   scale_x_continuous(breaks = c(seq(0, 80, by = 20), 200, 400, 600),
                      labels = c(seq(0, 80, by = 20), 200, 400, 600)) +
   scale_y_continuous(breaks = c(0.1, seq(0, 1, by = 0.2)),
                      labels = c(0.1, seq(0, 1, by = 0.2))) +
   annotate("text", y = 0.2, x = 300, label="10% of the sampled bp have \nat least 40 overlapping reads",
            hjust=0.5, vjust=1, size = 6)


# READ DEPTH DATA GENERATED BY SAMTOOLS DEPTH
dep <- read.table(file = "Bioinformatics_pipeline_development_task/Read_coverage/chr19_depth.txt")
head(dep)
tail(dep)
colnames(dep) <- c("CHROM", "POS", "DEPTH")
range(dep$POS)

table(dep$DEPTH>0)
3497613/length(min(dep$POS):max(dep$POS))

head(dep[dep$DEPTH==0,])
sum(dep$DEPTH)
dep[dep$DEPTH==max(dep$DEPTH),]

deeptT <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_coverage.txt")
head(deeptT)
range(deeptT$V2)
range(deeptT$V3)
table(deeptT$V4>0)
249777/length(min(dep$POS):max(dep$POS))
mean(deeptT$V4)
deeptT[deeptT$V4==max(deeptT$V4),]
deeptT[6900:6950,]

# PLOT DEPTH ACROSS ENTIRE CHROMOSOME 19
range(dep$DEPTH)
plot(dep$DEPTH,
     type='h',
     ylim=c(0,1100),
     col='blue',
     ylab='Coverage',
     xlab='Position',
     xaxt='n',
     main='Coverage at chr19'
)

# GENERATE X-AXIS
l <- length(dep$POS)
axis(1,
     at = seq(from = 0, to = l, length.out = 18),
     labels = seq(from = head(dep$POS,n=1)-1,
                  to=tail(dep$POS, n=1),
                  by = l/4
     )
)

mean(dep$DEPTH)
bedgr <- read.table("Bioinformatics_pipeline_development_task/results/1_read_coverage/chr19_coverage.bedgraph")
head(bedgr)
mean(bedgr$V4)

deps <- dep[sample(x = 1:nrow(dep), size = 1000, replace = FALSE), ]
head(deps)
ggplot(data = deps) +
   geom_segment(aes(x = POS, y = 0, xend = POS, yend = DEPTH))

# READ TARGET FILE
tar <- read.table(file = "Files_needed_for_task/target_regions.bed")
head(tar)
colnames(tar) <- c("CHROM", "START", "END")
range(tar$START)
range(tar$END)

#GENOME FRACTION COVERAGE
fra <- read.table("Read_coverage/raw_data_qualimapReport/genome_fraction_coverage.txt")
head(fra)
sum(fra$V2)
